src/main/java/net/minecraftforge/common/world/RegistryAccessExtension.java

```diff
@@ -0,0 +1,123 @@
+
+/**
+ * Defines an additional world-gen component to load as part of the RegistryAccess registries.
+ * <p>
+ * The path of the provided ResourceKey becomes the data folder from which datapack definitions are parsed
+ * and the provided Codec handles encoding and decoding to and from the data format.
+ * <p>
+ * To avoid conflicts with other mods, the path of the ResourceKey must be prefixed with the owning mod's
+ * namespace (ie {@code <modid>:<modid>/path}). Static helper methods for creating valid ResourceKeys are
+ * provided at {@link #createRegistryKey(String, String)} and {@link #createRegistryKey(ResourceLocation)}.
+ * <p>
+ * Default or 'built-in' registry entries can be registered via a forge DeferredRegister or IForgeRegistry
+ * of the same registry ResourceKey. These elements will populate the resulting RegistryAccess before datapacks
+ * are loaded and may be overridden by data-definitions. A static helper method for creating a DeferredRegisters
+ * for a RegistryAccessExtension is provided at {@link #createRegister(Class, ResourceKey)}.
+ *
+ * @param <T> The type of world-gen component.
+ */
+public class RegistryAccessExtension<T extends IForgeRegistryEntry<T>> extends ForgeRegistryEntry<RegistryAccessExtension<?>>
+{
+    private final ResourceKey<Registry<T>> registryKey;
+    private final Codec<T> directCodec;
+    private final Lifecycle defaultElementLifecycle;
+
+    public RegistryAccessExtension(ResourceKey<Registry<T>> registryKey, Codec<T> directCodec)
+    {
+        this(registryKey, directCodec, Lifecycle.experimental());
+    }
+
+    public RegistryAccessExtension(ResourceKey<Registry<T>> registryKey, Codec<T> directCodec, Lifecycle defaultElementLifecycle)
+    {
+        validateRegistryKey(registryKey);
+        setRegistryName(registryKey.location());
+        this.registryKey = registryKey;
+        this.directCodec = directCodec;
+        this.defaultElementLifecycle = defaultElementLifecycle;
+    }
+
+    public ResourceKey<? extends Registry<T>> getRegistryKey()
+    {
+        return registryKey;
+    }
+
+    public Codec<T> getDirectCodec()
+    {
+        return directCodec;
+    }
+
+    public Lifecycle getDefaultElementLifecycle()
+    {
+        return defaultElementLifecycle;
+    }
+
+    public static <T> ResourceKey<Registry<T>> createRegistryKey(String namespace, String path)
+    {
+        var prefix = getPathPrefix(namespace);
+        if (!path.startsWith(prefix))
+        {
+            path = prefix + path;
+        }
+
+        return ResourceKey.createRegistryKey(new ResourceLocation(namespace, path));
+    }
+
+    public static <T> ResourceKey<Registry<T>> createRegistryKey(ResourceLocation location)
+    {
+        return createRegistryKey(location.getNamespace(), location.getPath());
+    }
+
+    public static <T extends IForgeRegistryEntry<T>> DeferredRegister<T> createRegister(Class<T> type, ResourceKey<Registry<T>> key)
```

Comment:
Why couldn't this be an instance method since the data already exists on the extension access? It also seems as though this is responsible for creating both the registry and the entries. I think it would be better to split the deferred register into a creation event for the static registry if wanted and then an instance method to create the deferred register for the specified mod id.

Reply:
I'm not sure if misunderstanding, but is it not a similar chicken-egg situation as the above point where you don't want RegAccExtns statically initialized, but they'd kindof have to be for component deferred registers to be created off instance methods of them?

Could you reword please if I've missed the point?

Reply:
Apologies, I think I forgot the order I commented against the last time. You can resolve this comment.